lib/runner/variant-reassignment.js

```diff
@@ -212,14 +216,96 @@ export default class VariantReassignment {
     return true
   }
 
+  /**
+   * Variants will be removed from a product for 2 reasons:
+   * 1) variants will be moved to a CTP product to update from matching products
+   * 2) variants needs to be removed from CTP product to update because they don't exist
+   * in the new product draft anymore
+   *
+   * @param productDraft
+   * @param matchingProducts
+   * @param ctpProductToUpdate
+   * @returns {{matchingProducts, ctpProductToUpdate}}
+   * @private
+   */
+  _getRemovedVariants (productDraft, matchingProducts, ctpProductToUpdate) {
```

Comment:
This function looks complicated. I tried to modify `_getProductVariants` so it returns a map of variants indexed by SKU instead of a plain array and now it is possible to simplify `_getRemovedVariants` to this:
```js
import _ from 'lodash'

class Test {
  _getRemovedVariants (productDraft, matchingProducts, ctpProductToUpdate) {
    const skus = this._getProductDraftSkus(productDraft)
    const productsToRemoveVariants = matchingProducts.filter(p => p !== ctpProductToUpdate)

    // variants which will be moved to ctpProductToUpdate
    const matchingProductsVars = productsToRemoveVariants.reduce((variants, product) => {
      // get all variants from a product in object indexed by SKU
      const variantsMap = this._getProductVariantsMapBySku(product)
      // pick only variants which we want to move and merge with map of all variants of 
      return _.merge(variants, _.pick(variantsMap, skus))
    }, {})
    
    // variants which will be removed from ctpProductToUpdate
    const ctpProductToUpdateVars = this._getProductVariantsMapBySku(ctpProductToUpdate)
    // delete from ctpProductToUpdateVars map variants which should be preserved in ctpProductToUpdate
    skus.forEach((sku) => (delete ctpProductToUpdateVars[sku]))

    // return objects with variants indexed by SKUs
    return { matchingProductsVars, ctpProductToUpdateVars }
  }

  _getProductVariantsMapBySku (product) {
    const { current, staged } = product.masterData
    const variants = {
      ..._.keyBy(current.variants.concat(current.masterVariant), 'sku'),
      ..._.keyBy(staged.variants.concat(staged.masterVariant), 'sku')
    }
  }
  
  _getProductDraftSkus (product) {
    const variants = [product.masterVariant].concat(product.variants || [])
    return _.map(variants, 'sku')
  }
}
```

Maybe you can use something from it so it gets less complicated.

Reply:
https://github.com/commercetools/commercetools-node-variant-reassignment/pull/23/commits/e4cd1fc5c7b1b9bcc9b3ee41dd90f9e1e2d711c8